From a8a6faba53af7e0a8bd1d204c871d8a84fd77c18 Mon Sep 17 00:00:00 2001
From: Zhang Zekun <zhangzekun11@huawei.com>
Date: Mon, 10 Oct 2022 12:40:11 +0000
Subject: [PATCH 0504/1076] [Huawei] add config to control the export of
 collect_procs()

Offering: HULK
hulk inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I53VVE

------------------------------------------

Add a config to control the feature in bugzilla which
export the function collect_procs().

Signed-off-by: Zhang Zekun <zhangzekun11@huawei.com>
---
 arch/arm64/configs/openeuler_defconfig | 1 +
 arch/x86/configs/openeuler_defconfig   | 1 +
 include/linux/mm.h                     | 3 +++
 mm/Kconfig                             | 6 ++++++
 mm/memory-failure.c                    | 7 +++++++
 5 files changed, 18 insertions(+)

diff --git a/arch/arm64/configs/openeuler_defconfig b/arch/arm64/configs/openeuler_defconfig
index 737ad58ebf87..4c23df4afee2 100644
--- a/arch/arm64/configs/openeuler_defconfig
+++ b/arch/arm64/configs/openeuler_defconfig
@@ -1112,6 +1112,7 @@ CONFIG_ARCH_HAS_PTE_SPECIAL=y
 CONFIG_PIN_MEMORY=y
 CONFIG_PID_RESERVE=y
 CONFIG_MEMORY_RELIABLE=y
+CONFIG_EXPORT_COLLECT_PROCS=y
 
 #
 # Data Access Monitoring
diff --git a/arch/x86/configs/openeuler_defconfig b/arch/x86/configs/openeuler_defconfig
index c005578a2e02..40f6d273147c 100644
--- a/arch/x86/configs/openeuler_defconfig
+++ b/arch/x86/configs/openeuler_defconfig
@@ -1062,6 +1062,7 @@ CONFIG_ARCH_HAS_PKEYS=y
 # CONFIG_READ_ONLY_THP_FOR_FS is not set
 CONFIG_ARCH_HAS_PTE_SPECIAL=y
 CONFIG_MAPPING_DIRTY_HELPERS=y
+CONFIG_EXPORT_COLLECT_PROCS=y
 
 #
 # Data Access Monitoring
diff --git a/include/linux/mm.h b/include/linux/mm.h
index 772731651e9d..2f97855321e7 100644
--- a/include/linux/mm.h
+++ b/include/linux/mm.h
@@ -3108,8 +3108,11 @@ extern int sysctl_memory_failure_recovery;
 extern void shake_page(struct page *p, int access);
 extern atomic_long_t num_poisoned_pages __read_mostly;
 extern int soft_offline_page(unsigned long pfn, int flags);
+
+#ifdef CONFIG_EXPORT_COLLECT_PROCS
 extern void collect_procs(struct page *page, struct list_head *tokill,
 				int force_early);
+#endif
 
 /*
  * Error handlers for various types of pages.
diff --git a/mm/Kconfig b/mm/Kconfig
index a0a71d95e1ad..355a858174c4 100644
--- a/mm/Kconfig
+++ b/mm/Kconfig
@@ -992,6 +992,12 @@ config CLEAR_FREELIST_PAGE
 	  To enable this feature, kernel parameter "clear_freelist" also
 	  needs to be added.
 
+config EXPORT_COLLECT_PROCS
+       bool "Export the function collect_procs()"
+       default n
+       help
+         Export this function to collect the processes who have the page
+         mapped via collect_procs().
 
 source "mm/damon/Kconfig"
 
diff --git a/mm/memory-failure.c b/mm/memory-failure.c
index be29618db2f5..eefd84154ba6 100644
--- a/mm/memory-failure.c
+++ b/mm/memory-failure.c
@@ -545,8 +545,13 @@ static void collect_procs_file(struct page *page, struct list_head *to_kill,
 /*
  * Collect the processes who have the corrupted page mapped to kill.
  */
+#ifdef CONFIG_EXPORT_COLLECT_PROCS
 void collect_procs(struct page *page, struct list_head *tokill,
 				int force_early)
+#else
+static void collect_procs(struct page *page, struct list_head *tokill,
+				int force_early)
+#endif
 {
 	if (!page->mapping)
 		return;
@@ -556,7 +561,9 @@ void collect_procs(struct page *page, struct list_head *tokill,
 	else
 		collect_procs_file(page, tokill, force_early);
 }
+#ifdef CONFIG_EXPORT_COLLECT_PROCS
 EXPORT_SYMBOL_GPL(collect_procs);
+#endif
 
 static const char *action_name[] = {
 	[MF_IGNORED] = "Ignored",
-- 
2.17.1

